Fix linking Mac Catalyst by including LC_BUILD_VERSION in object files

bmisiak opened the pull request:
Hello.

Issue #106021 prevents Rust code from being linked into Mac Catalyst applications. Apple's LD has started requiring object files to contain version information about the platform they were built for, such as:
* the "deployment target" (minimum supported OS version), 
* the SDK version
* the type of the platform (macOS/iOS/catalyst/tvOS/watchOS all have a different number). 

This is currently only enforced when building for Mac Catalyst.

Rust uses the `object` crate which added support for including this information starting with `0.31.0`. ~~I upgraded it along with `thorin-dwp` so that everything depends on 0.31.
Apparently 0.31 [pulls in](https://github.com/gimli-rs/object/issues/463) `ruzstd` due to a [new ELF standard](https://maskray.me/blog/2022-09-09-zstd-compressed-debug-sections) because its `compression` feature is enabled by thorin. If you find this objectionable, let me know what the best way to avoid pulling in those dependencies might be.~~

**(`object` upgraded in https://github.com/rust-lang/rust/pull/111413)**

I then added two commits:
* The first one adds very basic, hard-coded support for calling `set_macho_build_version` for `-macabi` (Catalyst) targets, where it claims deployment target of Catalyst 14.0 and SDK of 16.2.
* The second weaves the versioning through `rust_target::spec::TargetOptions`, so that we can stick to specifying all target-related info in one place.

Kudos to @ara4n for writing [this gist](https://gist.github.com/ara4n/320a53ea768aba51afad4c9ed2168536).

bmisiak committed:
Fix Mac Catalyst linking by adding build version
Issue #106021
Apple LD requires build versions in object files for Catalyst

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4933560796/jobs/8817632188) [(plain)](https://github.com/rust-lang/rust/commit/9012e506a6f247246e0ebafdf74544ca5d8a0879/checks/13364237732/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:8e5e7e5ab8b370d6c329ec480221332ada57f0ab)
Download action repository 'rust-lang/simpleinfra@master' (SHA:af9a3f1b7e12a54c737d8aa371acc8d05cb83a8f)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180119 sha256=9fa76c45f3193f307e02ea67d1a48cfe7ef2b854a074b766938a88e046bc7887
  Stored in directory: /tmp/pip-ephem-wheel-cache-nqoq4m62/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully tagged rust-ci:latest
Built container sha256:0b7e795758233f79cae90f3707bd788b8c721de21e32979ba0287ef2d797bce9
Uploading finished image to https://ci-caches.rust-lang.org/docker/e51156f19850ce886cec818e46dc2a021e0aa7c270d15673e8fe74cd8522fc8ac3995109aebb688ee49ed586735ed4cf5f8c06d44c48298fa09c35ae4b082281

<botocore.awsrequest.AWSRequest object at 0x7ff20e9dffd0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 16.64s
##[endgroup]
fmt check
Diff in /checkout/compiler/rustc_target/src/spec/x86_64_apple_ios_macabi.rs at line 1:
 use super::apple_base::{
-    Arch,
-    mac_catalyst_deployment_target,
-    mac_catalyst_llvm_target,
-    mac_catalyst_sdk_version,
-    opts,
+    mac_catalyst_deployment_target, mac_catalyst_llvm_target, mac_catalyst_sdk_version, opts, Arch,
 };
 use crate::spec::{Cc, LinkerFlavor, Lld, StackProbeType, Target, TargetOptions};
 
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/aarch64_be_unknown_linux_gnu_ilp32.rs" "/checkout/compiler/rustc_target/src/spec/armv5te_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/armv7a_kmc_solid_asp3_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/i686_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/aarch64_pc_windows_gnullvm.rs" "/checkout/compiler/rustc_target/src/spec/sparc64_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/asmjs_unknown_emscripten.rs" "/checkout/compiler/rustc_target/src/spec/x86_64_apple_ios_macabi.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:26
Diff in /checkout/compiler/rustc_target/src/spec/aarch64_apple_ios_macabi.rs at line 1:
 use super::apple_base::{
-    Arch,
-    mac_catalyst_deployment_target,
-    mac_catalyst_llvm_target,
-    mac_catalyst_sdk_version,
-    opts,
+    mac_catalyst_deployment_target, mac_catalyst_llvm_target, mac_catalyst_sdk_version, opts, Arch,
 };
 use crate::spec::{Cc, FramePointer, LinkerFlavor, Lld, Target, TargetOptions};
```

</details>

workingjubilee reviewed (commented):
It would be best to also bump object in `rustc_codegen_llvm` as otherwise rustc will be using two different versions of the object crate during compilation which could have unfortunate effects.

It is not necessary to change std's dependencies here because std needs backtrace to be updated first.

I have opened #111413 which does this but otherwise has some conflicting changes with your PR, so I'm quite happy to drop mine.

petrochenkov commented:
Blocked on https://github.com/rust-lang/rust/pull/111413.

bmisiak commented:
@workingjubilee I did bump it here as well, happy to proceed either way. 👌

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4938695352/jobs/8828743220) [(plain)](https://github.com/rust-lang/rust/commit/df5cd56b5b61ca268b5a589626cac7ea614ccf9d/checks/13378619198/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......F..............................

failures:

---- spec::tests::aarch64_apple_ios_macabi stdout ----
thread 'spec::tests::aarch64_apple_ios_macabi' panicked at 'assertion failed: `(left == right)`
  left: `Ok(Target { llvm_target: "arm64-apple-ios14.0-macabi", pointer_width: 64, arch: "aarch64", data_layout: "e-m:o-i64:64-i128:128-n32:64-S128", options: TargetOptions { is_builtin: false, endian: little, c_int_width: "32", os: "ios", env: "", abi: "macabi", vendor: "apple", linker: None, linker_flavor: Darwin(Yes, No), linker_flavor_json: Gcc, lld_flavor_json: Ld64, linker_is_gnu_json: false, pre_link_objects: {}, post_link_objects: {}, pre_link_objects_self_contained: {}, post_link_objects_self_contained: {}, link_self_contained: False, pre_link_args: {Darwin(Yes, Yes): ["-target", "arm64-apple-ios14.0-macabi"], Darwin(Yes, No): ["-target", "arm64-apple-ios14.0-macabi"], Darwin(No, Yes): ["-arch", "arm64", "-platform_version", "mac-catalyst", "7.0", "7.0"], Darwin(No, No): ["-arch", "arm64", "-platform_version", "mac-catalyst", "7.0", "7.0"]}, pre_link_args_json: {Gcc: ["-target", "arm64-apple-ios14.0-macabi"], Ld: ["-arch", "arm64", "-platform_version", "mac-catalyst", "7.0", "7.0"], Lld(Ld64): ["-arch", "arm64", "-platform_version", "mac-catalyst", "7.0", "7.0"]}, late_link_args: {}, late_link_args_json: {}, late_link_args_dynamic: {}, late_link_args_dynamic_json: {}, late_link_args_static: {}, late_link_args_static_json: {}, post_link_args: {}, post_link_args_json: {}, link_script: None, link_env: [("ZERO_AR_DATE", "1")], link_env_remove: ["IPHONEOS_DEPLOYMENT_TARGET"], asm_args: [], cpu: "apple-a12", features: "+neon,+fp-armv8,+apple-a12", dynamic_linking: true, dll_tls_export: true, only_cdylib: false, executables: true, relocation_model: Pic, code_model: None, tls_model: GeneralDynamic, disable_redzone: false, frame_pointer: NonLeaf, function_sections: false, dll_prefix: "lib", dll_suffix: ".dylib", exe_suffix: "", staticlib_prefix: "lib", staticlib_suffix: ".a", families: ["unix"], abi_return_struct_as_int: true, is_like_aix: false, is_like_osx: true, is_like_solaris: false, is_like_windows: false, is_like_msvc: false, is_like_wasm: false, is_like_android: false, default_dwarf_version: 2, allows_weak_linkage: true, has_rpath: true, no_default_libraries: true, position_independent_executables: false, static_position_independent_executables: false, needs_plt: false, relro_level: None, archive_format: "darwin", allow_asm: true, main_needs_argc_argv: true, has_thread_local: false, obj_is_bitcode: false, forces_embed_bitcode: true, bitcode_llvm_cmdline: "-triple\0arm64-apple-ios-macabi\0-emit-obj\0-disable-llvm-passes\0-Os\0", macho_build_version_min_os: "", macho_build_version_sdk: "", macho_build_version_platform: "", min_atomic_width: None, max_atomic_width: Some(128), atomic_cas: true, panic_strategy: Unwind, crt_static_allows_dylibs: false, crt_static_default: false, crt_static_respected: false, stack_probes: None, min_global_align: None, default_codegen_units: None, trap_unreachable: true, requires_lto: false, singlethread: false, no_builtins: false, default_hidden_visibility: false, emit_debug_gdb_scripts: false, requires_uwtable: false, default_uwtable: false, simd_types_indirect: true, limit_rdylib_exports: true, override_export_symbols: None, merge_functions: Aliases, mcount: "mcount", llvm_abiname: "", relax_elf_relocations: false, llvm_args: [], use_ctors_section: false, eh_frame_header: false, has_thumb_interworking: false, debuginfo_kind: DwarfDsym, split_debuginfo: Packed, supported_split_debuginfo: [Packed, Unpacked, Off], supported_sanitizers: (empty), default_adjusted_cabi: None, c_enum_min_bits: None, generate_arange_section: true, supports_stack_protector: true, entry_name: "main", entry_abi: C, supports_xray: false, force_emulated_tls: false } })`,
 right: `Ok(Target { llvm_target: "arm64-apple-ios14.0-macabi", pointer_width: 64, arch: "aarch64", data_layout: "e-m:o-i64:64-i128:128-n32:64-S128", options: TargetOptions { is_builtin: false, endian: little, c_int_width: "32", os: "ios", env: "", abi: "macabi", vendor: "apple", linker: None, linker_flavor: Darwin(Yes, No), linker_flavor_json: Gcc, lld_flavor_json: Ld64, linker_is_gnu_json: false, pre_link_objects: {}, post_link_objects: {}, pre_link_objects_self_contained: {}, post_link_objects_self_contained: {}, link_self_contained: False, pre_link_args: {Darwin(Yes, Yes): ["-target", "arm64-apple-ios14.0-macabi"], Darwin(Yes, No): ["-target", "arm64-apple-ios14.0-macabi"], Darwin(No, Yes): ["-arch", "arm64", "-platform_version", "mac-catalyst", "7.0", "7.0"], Darwin(No, No): ["-arch", "arm64", "-platform_version", "mac-catalyst", "7.0", "7.0"]}, pre_link_args_json: {Gcc: ["-target", "arm64-apple-ios14.0-macabi"], Ld: ["-arch", "arm64", "-platform_version", "mac-catalyst", "7.0", "7.0"], Lld(Ld64): ["-arch", "arm64", "-platform_version", "mac-catalyst", "7.0", "7.0"]}, late_link_args: {}, late_link_args_json: {}, late_link_args_dynamic: {}, late_link_args_dynamic_json: {}, late_link_args_static: {}, late_link_args_static_json: {}, post_link_args: {}, post_link_args_json: {}, link_script: None, link_env: [("ZERO_AR_DATE", "1")], link_env_remove: ["IPHONEOS_DEPLOYMENT_TARGET"], asm_args: [], cpu: "apple-a12", features: "+neon,+fp-armv8,+apple-a12", dynamic_linking: true, dll_tls_export: true, only_cdylib: false, executables: true, relocation_model: Pic, code_model: None, tls_model: GeneralDynamic, disable_redzone: false, frame_pointer: NonLeaf, function_sections: false, dll_prefix: "lib", dll_suffix: ".dylib", exe_suffix: "", staticlib_prefix: "lib", staticlib_suffix: ".a", families: ["unix"], abi_return_struct_as_int: true, is_like_aix: false, is_like_osx: true, is_like_solaris: false, is_like_windows: false, is_like_msvc: false, is_like_wasm: false, is_like_android: false, default_dwarf_version: 2, allows_weak_linkage: true, has_rpath: true, no_default_libraries: true, position_independent_executables: false, static_position_independent_executables: false, needs_plt: false, relro_level: None, archive_format: "darwin", allow_asm: true, main_needs_argc_argv: true, has_thread_local: false, obj_is_bitcode: false, forces_embed_bitcode: true, bitcode_llvm_cmdline: "-triple\0arm64-apple-ios-macabi\0-emit-obj\0-disable-llvm-passes\0-Os\0", macho_build_version_min_os: "14.0", macho_build_version_sdk: "16.2", macho_build_version_platform: "maccatalyst", min_atomic_width: None, max_atomic_width: Some(128), atomic_cas: true, panic_strategy: Unwind, crt_static_allows_dylibs: false, crt_static_default: false, crt_static_respected: false, stack_probes: None, min_global_align: None, default_codegen_units: None, trap_unreachable: true, requires_lto: false, singlethread: false, no_builtins: false, default_hidden_visibility: false, emit_debug_gdb_scripts: false, requires_uwtable: false, default_uwtable: false, simd_types_indirect: true, limit_rdylib_exports: true, override_export_symbols: None, merge_functions: Aliases, mcount: "mcount", llvm_abiname: "", relax_elf_relocations: false, llvm_args: [], use_ctors_section: false, eh_frame_header: false, has_thumb_interworking: false, debuginfo_kind: DwarfDsym, split_debuginfo: Packed, supported_split_debuginfo: [Packed, Unpacked, Off], supported_sanitizers: (empty), default_adjusted_cabi: None, c_enum_min_bits: None, generate_arange_section: true, supports_stack_protector: true, entry_name: "main", entry_abi: C, supports_xray: false, force_emulated_tls: false } })`', compiler/rustc_target/src/spec/tests/tests_impl.rs:9:5

---- spec::tests::x86_64_apple_ios_macabi stdout ----
---- spec::tests::x86_64_apple_ios_macabi stdout ----
thread 'spec::tests::x86_64_apple_ios_macabi' panicked at 'assertion failed: `(left == right)`
error: test failed, to rerun pass `-p rustc_target --lib`
  left: `Ok(Target { llvm_target: "x86_64-apple-ios14.0-macabi", pointer_width: 64, arch: "x86_64", data_layout: "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", options: TargetOptions { is_builtin: false, endian: little, c_int_width: "32", os: "ios", env: "", abi: "macabi", vendor: "apple", linker: None, linker_flavor: Darwin(Yes, No), linker_flavor_json: Gcc, lld_flavor_json: Ld64, linker_is_gnu_json: false, pre_link_objects: {}, post_link_objects: {}, pre_link_objects_self_contained: {}, post_link_objects_self_contained: {}, link_self_contained: False, pre_link_args: {Darwin(Yes, Yes): ["-target", "x86_64-apple-ios14.0-macabi"], Darwin(Yes, No): ["-target", "x86_64-apple-ios14.0-macabi"], Darwin(No, Yes): ["-arch", "x86_64", "-platform_version", "mac-catalyst", "7.0", "7.0"], Darwin(No, No): ["-arch", "x86_64", "-platform_version", "mac-catalyst", "7.0", "7.0"]}, pre_link_args_json: {Gcc: ["-target", "x86_64-apple-ios14.0-macabi"], Ld: ["-arch", "x86_64", "-platform_version", "mac-catalyst", "7.0", "7.0"], Lld(Ld64): ["-arch", "x86_64", "-platform_version", "mac-catalyst", "7.0", "7.0"]}, late_link_args: {}, late_link_args_json: {}, late_link_args_dynamic: {}, late_link_args_dynamic_json: {}, late_link_args_static: {}, late_link_args_static_json: {}, post_link_args: {}, post_link_args_json: {}, link_script: None, link_env: [("ZERO_AR_DATE", "1")], link_env_remove: ["IPHONEOS_DEPLOYMENT_TARGET"], asm_args: [], cpu: "core2", features: "", dynamic_linking: true, dll_tls_export: true, only_cdylib: false, executables: true, relocation_model: Pic, code_model: None, tls_model: GeneralDynamic, disable_redzone: false, frame_pointer: Always, function_sections: false, dll_prefix: "lib", dll_suffix: ".dylib", exe_suffix: "", staticlib_prefix: "lib", staticlib_suffix: ".a", families: ["unix"], abi_return_struct_as_int: true, is_like_aix: false, is_like_osx: true, is_like_solaris: false, is_like_windows: false, is_like_msvc: false, is_like_wasm: false, is_like_android: false, default_dwarf_version: 2, allows_weak_linkage: true, has_rpath: true, no_default_libraries: true, position_independent_executables: false, static_position_independent_executables: false, needs_plt: false, relro_level: None, archive_format: "darwin", allow_asm: true, main_needs_argc_argv: true, has_thread_local: false, obj_is_bitcode: false, forces_embed_bitcode: false, bitcode_llvm_cmdline: "", macho_build_version_min_os: "", macho_build_version_sdk: "", macho_build_version_platform: "", min_atomic_width: None, max_atomic_width: Some(64), atomic_cas: true, panic_strategy: Unwind, crt_static_allows_dylibs: false, crt_static_default: false, crt_static_respected: false, stack_probes: InlineOrCall { min_llvm_version_for_inline: (16, 0, 0) }, min_global_align: None, default_codegen_units: None, trap_unreachable: true, requires_lto: false, singlethread: false, no_builtins: false, default_hidden_visibility: false, emit_debug_gdb_scripts: false, requires_uwtable: false, default_uwtable: false, simd_types_indirect: true, limit_rdylib_exports: true, override_export_symbols: None, merge_functions: Aliases, mcount: "mcount", llvm_abiname: "", relax_elf_relocations: false, llvm_args: [], use_ctors_section: false, eh_frame_header: false, has_thumb_interworking: false, debuginfo_kind: DwarfDsym, split_debuginfo: Packed, supported_split_debuginfo: [Packed, Unpacked, Off], supported_sanitizers: (empty), default_adjusted_cabi: None, c_enum_min_bits: None, generate_arange_section: true, supports_stack_protector: true, entry_name: "main", entry_abi: C, supports_xray: false, force_emulated_tls: false } })`,
 right: `Ok(Target { llvm_target: "x86_64-apple-ios14.0-macabi", pointer_width: 64, arch: "x86_64", data_layout: "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", options: TargetOptions { is_builtin: false, endian: little, c_int_width: "32", os: "ios", env: "", abi: "macabi", vendor: "apple", linker: None, linker_flavor: Darwin(Yes, No), linker_flavor_json: Gcc, lld_flavor_json: Ld64, linker_is_gnu_json: false, pre_link_objects: {}, post_link_objects: {}, pre_link_objects_self_contained: {}, post_link_objects_self_contained: {}, link_self_contained: False, pre_link_args: {Darwin(Yes, Yes): ["-target", "x86_64-apple-ios14.0-macabi"], Darwin(Yes, No): ["-target", "x86_64-apple-ios14.0-macabi"], Darwin(No, Yes): ["-arch", "x86_64", "-platform_version", "mac-catalyst", "7.0", "7.0"], Darwin(No, No): ["-arch", "x86_64", "-platform_version", "mac-catalyst", "7.0", "7.0"]}, pre_link_args_json: {Gcc: ["-target", "x86_64-apple-ios14.0-macabi"], Ld: ["-arch", "x86_64", "-platform_version", "mac-catalyst", "7.0", "7.0"], Lld(Ld64): ["-arch", "x86_64", "-platform_version", "mac-catalyst", "7.0", "7.0"]}, late_link_args: {}, late_link_args_json: {}, late_link_args_dynamic: {}, late_link_args_dynamic_json: {}, late_link_args_static: {}, late_link_args_static_json: {}, post_link_args: {}, post_link_args_json: {}, link_script: None, link_env: [("ZERO_AR_DATE", "1")], link_env_remove: ["IPHONEOS_DEPLOYMENT_TARGET"], asm_args: [], cpu: "core2", features: "", dynamic_linking: true, dll_tls_export: true, only_cdylib: false, executables: true, relocation_model: Pic, code_model: None, tls_model: GeneralDynamic, disable_redzone: false, frame_pointer: Always, function_sections: false, dll_prefix: "lib", dll_suffix: ".dylib", exe_suffix: "", staticlib_prefix: "lib", staticlib_suffix: ".a", families: ["unix"], abi_return_struct_as_int: true, is_like_aix: false, is_like_osx: true, is_like_solaris: false, is_like_windows: false, is_like_msvc: false, is_like_wasm: false, is_like_android: false, default_dwarf_version: 2, allows_weak_linkage: true, has_rpath: true, no_default_libraries: true, position_independent_executables: false, static_position_independent_executables: false, needs_plt: false, relro_level: None, archive_format: "darwin", allow_asm: true, main_needs_argc_argv: true, has_thread_local: false, obj_is_bitcode: false, forces_embed_bitcode: false, bitcode_llvm_cmdline: "", macho_build_version_min_os: "14.0", macho_build_version_sdk: "16.2", macho_build_version_platform: "maccatalyst", min_atomic_width: None, max_atomic_width: Some(64), atomic_cas: true, panic_strategy: Unwind, crt_static_allows_dylibs: false, crt_static_default: false, crt_static_respected: false, stack_probes: InlineOrCall { min_llvm_version_for_inline: (16, 0, 0) }, min_global_align: None, default_codegen_units: None, trap_unreachable: true, requires_lto: false, singlethread: false, no_builtins: false, default_hidden_visibility: false, emit_debug_gdb_scripts: false, requires_uwtable: false, default_uwtable: false, simd_types_indirect: true, limit_rdylib_exports: true, override_export_symbols: None, merge_functions: Aliases, mcount: "mcount", llvm_abiname: "", relax_elf_relocations: false, llvm_args: [], use_ctors_section: false, eh_frame_header: false, has_thumb_interworking: false, debuginfo_kind: DwarfDsym, split_debuginfo: Packed, supported_split_debuginfo: [Packed, Unpacked, Off], supported_sanitizers: (empty), default_adjusted_cabi: None, c_enum_min_bits: None, generate_arange_section: true, supports_stack_protector: true, entry_name: "main", entry_abi: C, supports_xray: false, force_emulated_tls: false } })`', compiler/rustc_target/src/spec/tests/tests_impl.rs:9:5

failures:
    spec::tests::aarch64_apple_ios_macabi
    spec::tests::x86_64_apple_ios_macabi
```

</details>

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -140,6 +140,9 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
     };
 
     let mut file = write::Object::new(binary_format, architecture, endianness);
+    if let Some(build_version) = macho_object_build_version_for_taget(&sess.target) {
```
```suggestion
    if let Some(build_version) = macho_object_build_version_for_target(&sess.target) {
```

workingjubilee reviewed (commented):
typo

ara4n commented:
(thanks for picking up my gist @bmisiak!)

bors commented:
:umbrella: The latest upstream changes (presumably #111413) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bmisiak commented on compiler/rustc_codegen_ssa/Cargo.toml:
```diff
@@ -7,7 +7,7 @@ edition = "2021"
 
 [dependencies]
-ar_archive_writer = "0.1.3"
+ar_archive_writer = "0.1.4"
```
bumps object dependency to 0.31 https://github.com/rust-lang/ar_archive_writer/pull/6
I believe it's already resolved to 0.1.4 as Cargo.lock seems to already have it

bmisiak commented:
I rebased it and removed the dependency changes except for the bump of `ar_archive_writer`, which already resolved to 0.1.4.

Two commits remain: the first one is a more basic support, the second one adds more complex changes to the Catalyst targets so that versioning can be specified in them.

@rustbot label -S-blocked +S-waiting-on-review

klensy commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -188,6 +188,9 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
     };
 
     let mut file = write::Object::new(binary_format, architecture, endianness);
+    if let Some(build_version) = macho_object_build_version_for_target(&sess.target) {
+        file.set_macho_build_version(build_version)
+    }
```
Can this be behind something like `sess.target.is_like_osx`?

imWildCat commented on compiler/rustc_target/src/spec/apple_base.rs:
```diff
@@ -284,6 +284,22 @@ fn ios_lld_platform_version() -> String {
+    // Seems that newer xcode refuses to compile for older than 14.0
+    // It returns a &'static str because TargetOptions::add_pre_link_args
+    // requires it.
+    if arch == "arm64" { "arm64-apple-ios14.0-macabi" } else { "x86_64-apple-ios14.0-macabi" }
```
For llvm, I believe we can use this triple without specifying target os version number.

For example, `arm64-apple-ios14.0-macabi` can be `arm64-apple-ios-macabi`.

```shell
$ clang -target x86_64-apple-ios-macabi demo.c # this can work
```

Similar case: #106925

imWildCat commented on compiler/rustc_target/src/spec/apple_base.rs:
```diff
@@ -284,6 +284,22 @@ fn ios_lld_platform_version() -> String {
 
+pub fn mac_catalyst_deployment_target() -> &'static str {
+    // Consider making this user-changeable in the future.
+    "14.0"
```
QQ: Could we re-use the value from environment variable `IPHONEOS_DEPLOYMENT_TARGET`?

I'm not sure about this. Found `IPHONEOS_DEPLOYMENT_TARGET` in the same file (apple_base.rs).

imWildCat commented on compiler/rustc_target/src/spec/apple_base.rs:
```diff
@@ -284,6 +284,22 @@ fn ios_lld_platform_version() -> String {
+    // Consider making this user-changeable in the future
+    "16.2"
+}
+pub fn mac_catalyst_llvm_target(arch: &str) -> &'static str {
```
Could we use a boolean or enum instead of `str`? 

There're only two cases for Mac Catalyst

klensy commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -258,6 +261,44 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
+        "iossim" => object::macho::PLATFORM_IOSSIMULATOR,
+        "tvossim" => object::macho::PLATFORM_TVOSSIMULATOR,
+        "watchossim" => object::macho::PLATFORM_WATCHOSSIMULATOR,
+        _ => return None,
```
Actually, if `sess.target.is_like_osx` == true, is that means that this return should be unreachable? Assuming if https://github.com/rust-lang/rust/pull/111384/files#r1199646361 applied.

petrochenkov commented:
In general target specs are sort of a public interface, so if some field is always `"14.0"` and no configurability is required right now, then I'd rather prefer hardcode in the style of the first commit.
Target specs could always be extended when it becomes actually necessary.
@rustbot author

petrochenkov commented:
Thanks!
@bors r+ rollup

bors commented:
:pushpin: Commit d816b8b26469db31cef8214a07756c3aa5009d79 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov d816b8b26469db31cef8214a07756c3aa5009d79 -->
<!-- homu: {"type":"Approved","sha":"d816b8b26469db31cef8214a07756c3aa5009d79","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bmisiak commented:
@petrochenkov Hmmm I'm finding that my std fails to build for Catalyst when `llvm_target` does not specify `ios14.0` instead of just `ios`. 

#106925 removed the version from llvm_target.

I might add it back in, bumping it from 13.0 (which broke on newer clang) to 14.0

bmisiak committed:
Mac Catalyst: specify 14.0 deployment taregt in llvm_target

bmisiak commented:
@rustbot label -S-waiting-on-bors +S-waiting-on-review

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit a61f026182cb0fa5830156b0ef29c28bc4ccfe99 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov a61f026182cb0fa5830156b0ef29c28bc4ccfe99 -->
<!-- homu: {"type":"Approved","sha":"a61f026182cb0fa5830156b0ef29c28bc4ccfe99","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
